gnuradio-runtime/lib/terminate_handler.cc

```diff
@@ -0,0 +1,137 @@
+#include <stdexcept>
+#include <iostream>
+#include <exception>
+#include <regex>
+#include <pmt/pmt.h>
+#include <boost/thread.hpp>
+
+#include <cxxabi.h>
+
+#ifdef HAVE_LIBUNWIND
+#define UNW_LOCAL_ONLY
+#include <libunwind.h>
+#include <cstdio>
+#include <cstdlib>
+#endif
+
+namespace gr {
+
+#ifdef HAVE_LIBUNWIND
+    void print_backtrace() {
+        unw_cursor_t cursor;
+        unw_context_t context;
+
+        // Initialize cursor to current frame for local unwinding.
+        unw_getcontext(&context);
+        unw_init_local(&cursor, &context);
+
+        // Unwind frames one by one, going up the frame stack.
+        while (unw_step(&cursor) > 0) {
+            unw_word_t offset, pc;
+            unw_get_reg(&cursor, UNW_REG_IP, &pc);
+            if (pc == 0) {
+                break;
+            }
+            std::fprintf(stderr, "0x%lx:", pc);
+
+            char sym[256];
+            if (unw_get_proc_name(&cursor, sym, sizeof(sym), &offset) == 0) {
+                char* nameptr = sym;
+                int status;
+                char* demangled = abi::__cxa_demangle(sym, nullptr, nullptr, &status);
+                if (status == 0) {
+                    nameptr = demangled;
+                }
+                std::fprintf(stderr, " (%s+0x%lx)\n", nameptr, offset);
+                std::free(demangled);
+            } else {
+                std::fprintf(stderr, " -- error: unable to obtain symbol name for this frame\n");
+            }
+        }
+    }
+#endif
+
+    void terminate_handler_impl() {
+        std::cerr << "terminate reached from: " << boost::this_thread::get_id();
```

Comment:
Adding "thread ID" here would make the error message more useful, since it may not be obvious what's getting printed after the colon.